lib/rosette/integrations/smartling-integration/smartling_puller.rb

```diff
@@ -1,135 +1,217 @@
 # encoding: UTF-8
 
+require 'set'
+require 'concurrent'
+
 module Rosette
   module Integrations
     class SmartlingIntegration < Integration
       class SmartlingPuller
 
-        attr_reader :rosette_config, :smartling_apis, :completed_files_map, :locale
+        DEFAULT_THREAD_POOL_SIZE = 15
+
+        attr_reader :rosette_config
+        attr_reader :repo_config, :serializer_id, :extractor_id
+        attr_reader :thread_pool_size, :logger
 
-        def initialize(rosette_config, smartling_apis, locale)
+        def initialize(rosette_config)
           @rosette_config = rosette_config
-          @smartling_apis = smartling_apis
-          @locale = locale
-          @completed_files_map = Hash.new { |h, key| h[key] = [] }
+          @thread_pool_size = DEFAULT_THREAD_POOL_SIZE
+          @logger = Rosette.logger
         end
 
-        def pull(repo_name, extractor_id)
-          each_file_for_repo(repo_name) do |file|
-            next unless file.repo_name == repo_name
-            next unless file_has_changed?(repo_name, file)
+        def set_repo_config(repo_config)
+          @repo_config = repo_config
+          self
+        end
 
-            potential_files = rosette_config.datastore.file_list_for_repo(repo_name)
+        def set_serializer_id(serializer_id)
```

Comment:
Yeah it's all about the chaining. All the `Command` classes in `Rosette::Core` follow the same pattern. I guess maybe I'm getting too Java up in here (this is pretty much how Java factories work too).